src/test/orphanage_tests.cpp

```diff
@@ -72,6 +72,341 @@ static bool EqualTxns(const std::set<CTransactionRef>& set_txns, const std::vect
+
+        // Number of peers = 3
+        orphanage->AddTx(txns.at(2), 2);
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS);
+        BOOST_CHECK_EQUAL(orphanage->ReservedPeerUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER);
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER * 3);
+        BOOST_CHECK_EQUAL(orphanage->MaxPeerAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS / 3);
+
+        // Number of peers didn't change.
+        orphanage->AddTx(txns.at(3), 2);
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS);
+        BOOST_CHECK_EQUAL(orphanage->ReservedPeerUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER);
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER * 3);
+        BOOST_CHECK_EQUAL(orphanage->MaxPeerAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS / 3);
+
+        // Once a peer has no orphans, it is not considered in the limits.
+        // Number of peers = 2
+        orphanage->EraseForPeer(2);
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS);
+        BOOST_CHECK_EQUAL(orphanage->ReservedPeerUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER);
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER * 2);
+        BOOST_CHECK_EQUAL(orphanage->MaxPeerAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS / 2);
+
+        // Number of peers = 1
+        orphanage->EraseTx(txns.at(0)->GetWitnessHash());
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS);
+        BOOST_CHECK_EQUAL(orphanage->ReservedPeerUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER);
+        BOOST_CHECK_EQUAL(orphanage->MaxGlobalUsage(), node::DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER);
+        BOOST_CHECK_EQUAL(orphanage->MaxPeerAnnouncements(), node::DEFAULT_MAX_ORPHAN_ANNOUNCEMENTS);
+    }
+
+    // Test eviction of multiple transactions at a time
+    {
+        // Create a large transaction that is 10 times larger than the normal size transaction.
+        CMutableTransaction tx_large;
+        tx_large.vin.resize(10);
+        tx_large.vout.resize(2);
+        for (int i = 0; i < 10; i++) {
+            tx_large.vin[i].prevout.n = i;
+            tx_large.vin[i].prevout.hash = Txid::FromUint256(m_rng.rand256());
+        }
+        const auto size_diff = 10 * TX_SIZE - GetTransactionWeight(*MakeTransactionRef(tx_large));
+        // Pad the last transaction until it is roughly 10 times larger than the normal size transaction.
+        tx_large.vout.back().scriptPubKey = CScript() << OP_RETURN << std::vector<unsigned char>(size_diff / WITNESS_SCALE_FACTOR);
```

Comment:
in 3e2985630cea8d59a6045e8a3451753130b18cab: could use the `BulkTransaction` helper here
```suggestion
        tx_large.vin.resize(1);
        BulkTransaction(tx_large, 10 * TX_SIZE);
```